### PR TITLE
Media migrations

### PR DESCRIPTION
**Pull request by emetsger**
## Testing
TL;DR: for instructions to test the PR [see below](#user-content-gh-desc-tests)

## Supported Media Migrations
This PR provides migration definitions and tests for migrating Media and Files into Drupal.  Migrations for the following Media types are provided:
* Generic files
* Images
* Video
* Remote Video (e.g. Youtube links)
* Extracted Text
* Audio
* Document

There is Media type for "FITS Technical metadata", which is _not_ supported by this PR.  It seems to be a special purpose Media type, and I'm not even sure if our repository is going to be accepting that kind of media, so I didn't spend time writing a migration definition and tests for it.  (Generally, I'm wondering of the Media types have been reviewed and approved by the stakeholders; some of the requirements - e.g. allowed file extensions - seem at best annoying and at worst will cause content creators to rename files to adapt to what appears as an arbitrary rule).

## Content-based Addressing of files
In addition to providing Media migration support, this PR supports content-addressed filenames for files that are ingested via migrations.  This means that files with identical byte content will have identical file names (same directory and filename).  Consequently, there will never be two different files in Drupal that have identical content.  A corollary is that it is impossible for a file to be accidentally overwritten in a migration process.

An important caveat is that content-based addressing of file content _**only applies to files provided in a migration**_.  The logic applied to interactively-uploaded content and to migrated content are separate and distinct from each other.  If a user creates a Media manually through the UI, the file associated with the Media will _**not**_ be content addressed, and therefore duplicates could be introduced if a duplicate file is uploaded manually.  However, if most Media are created by the migrations in this PR, then duplicates are not something to be too concerned with.  

Content-based addressing is achieved by introducing a new processing plugin named `pairtree`, and judicious use of various existing modules to provide more control over the filenames used when ingesting content during migration.  As an aside, SHA-1 is used on the file content, but if we feel it is insufficient we should be able to use any PHP-supported hashing algorithm.

If content addressing is not used, we will need to reason about the consequences of Drupal's current behavior, which is to match duplicates based on filename, regardless of their content, and navigate the various dials and knobs used to control its behavior (e.g. overwrite existing files with the duplicate, rename the duplicate until it is unique, etc).

## Treatment of `File` entities
This PR allows for multiple `File` entities to refer to the same file that is present on the Drupal filesystem.  

Take, for example, a migration that refers to a bytestream that has been previously migrated.  At one point in time, the content at http://cloudstorage.net/file/moo.pdf was migrated into Drupal, and at some later time, the same bytes are migrated again for some unknown reason; maybe a different URL was used the second time, but yielded the same bytes, who knows.

We know from content-based addressing that the second migration will not result in a duplicate file on the Drupal filesystem. 
 BUT, what do we do with the `File` entity that results?  This PR creates a new `File` entity for the second migration, instead of re-using the existing File entity from the original migration.  The rationale for creating a new `File` is that it allows a single bytestream to be known by distinct metadata (different values for the fields on a `File`) in multiple contexts.  The `File` may carry one `name` in one collection, and be known by another `name` in a different collection.   If the rationale doesn't hold water, we can easily change the behavior to re-use the existing `File` entity (matching on the content-addressed path).

`File` entities are cheap, they amount to a row in a database table, so from a resource perspective, the cost is negligible. Allowing potential "duplicate" entities does make accounting for orphan files (bytestreams) more difficult, and it is not clear what would happen if you were to delete a duplicate `File`: does deleting a duplicate `File` entity result in the bytestream being removed even though it is referenced by a different `File`?  That's unknown.  Anyway. The larger point is that this PR allows duplicate Files to be created and that it can be changed if we think that's problematic.

## Tests <a id="gh-desc-tests"></a>
The tests are developed in the same pattern as previous migrations (i.e. CSV spreadsheets imported by test cafe, expected resource content in JSON verified by the Go code), with this PR introducing a migration assets container that provides test media for ingest.  This keeps potentially large binaries used for testing out of the drupal image.

### Run the tests
Check out this PR
Run `make reset`
cd into the `tests` directory and execute `./10-migration-backend-tests.sh`; all the tests should succeed

cd into the `tests` directory and execute `./11-file-deletion-tests.sh`; all the tests should succeed

cd into the `tests` directory and execute `./12-migration-derivative-tests.sh`; all the tests should succeed

### Optionally
Run the tests above, and then log into Drupal, navigate to the repository object named `Media Migration Repository Item One` and view its Media by clicking on the "Media" label/tab.  Observe there are Media for each media type except `Remote Video` and `FITS Technical metadata`.  Feel free to click through the media and files there should be no surprises.

To observe content addressable storage and duplicate `File` entities, navigate to the repository object `Media Migration Repository Item Two` in a separate tab and view its Media.  Compare/contrast the Media named `Fuji Acros Datasheet` which appears in `Media Migration Repository Item One` and `Media Migration Repository Item Two`; their properties will be identical except for their member_of repository item.

Hover over the link labeled with the name of the Media: `Fuji Acros Datasheet`, and observe that for both entities the link is identical, and that the link references the checksum value `c9a060c39365820edc5d1a51f221d49e96a8a730` which is the checksum of the file associated with the `Fuji Acros Datasheet`.  Feel free to download the file and verify its SHA1 checksum.

Retrieve the JSON API representation of the `Fuji Acros Datasheet` media by visiting ` https://islandora-idc.traefik.me/jsonapi/media/document?filter[name]=Fuji Acros Datasheet` (you may need to encode the spaces in the query parameter with `%20` depending on your HTTP client).  There should be 2 objects in the `data` array, identical in most ways. 
Navigate to the object key `field_media_document`, for each of the results, and note that the file `id` is different for each.  This demonstrates that while each Media is using the same bytes, they have independent representation in the system as different `File` entities.  Verify this by navigating to the [Files tab](https://islandora-idc.traefik.me/admin/content/files), and in the list of files you should see the `Fuji Acros Datasheet` listed twice.

If you're a glutton for punishment, which you clearly are if you've gone this far, open `media-document.csv` in the `tests/10-migration-backend-tests/testcafe/migrations` directory.  Add a row to that spreadsheet, incrementing the local id and changing the value of the `name` and `original_name` fields to something arbitrary, leaving the rest of the fields as they appear.  In your browser, navigate to the [Migrate Source UI](https://islandora-idc.traefik.me/migrate_source_ui) and migrate these media by selecting "(3c) Ingest New Document Media..." and provide the edited `media-document.csv`as the source file.  Leave "Update Existing Records" checked.  Execute the migration.  Navigate/refresh the tab with `Media Migration Repository Item Two` in it, and observe your newly ingested Media.  It should have the same URI as the `Fuji Acros Datasheet` (the one that references checksum `c9a060c39365820edc5d1a51f221d49e96a8a730`).  Navigate to the [Files tab](https://islandora-idc.traefik.me/admin/content/files) and see your file listed, along with the two other `Fuji Acros Datasheet` files.

Finally, test out deletion behavior.  You cannot delete individual files, but you can "Delete Media and Associated Files" by navigating to the [Media Tab](https://islandora-idc.traefik.me/admin/content/media) and selecting one of the media named `Fuji Acros Datasheet` and select "Delete media and file(s)".  Navigate back to the [Files tab](https://islandora-idc.traefik.me/admin/content/files).  You may need to clear the Drupal cache at this point (click "Devel" on the top toolbar then click "Cache clear" on the second tool bar directly below the first), and refresh the [Files tab](https://islandora-idc.traefik.me/admin/content/files).  This should correct the counts in the "Used In" column.  Anyway, verify that one of the `Fuji Acros Datasheet` media have a "0" value for Used In.    Right click on hyperlinked name, and click save as.  Note that the file still downloads to your computer even though you deleted the Media and its associated files.  The `File` entity persists (it just has no references), and so does its bytestream (the file on disk in the Drupal filesystem).  This behavior is documented [here](https://github.com/jhu-idc/idc-isle-dc/issues/100) and in the `11-file-deletion-tests` test suite.

### CODE REVIEW

**Comment by birkland** [comment]
@emetsger This was a fascinating PR.  I think, reading the description, it makes sense to create issues for:
* review media types and suffix rules with stakeholders
* Verify consequences of File entities referring to same underlying bytestream (maybe something for vendor)
* Verify sufficiency of sha1 hashing

I also noticed a few UX issues that ought to at least be investigated in separate issues, unless you think it makes sense to address in this PR.   I'm not sure if stakeholders would care or not.
* Downloaded filenames do not have suffixes and correlate with the hash rather than filename from the metadata
* MIME type when downloading files always seems to be application/octet-stream
 
 As far as this relates to UX, this is what it looks like when trying to download an image in the browser (i.e. if I open it in a new tab, or click on the image)
 
image

**Review by birkland — APPROVED** [review]
This looks good  I noted in a comment above that we ought to create some github issues for examining some of the questions you raised in the description of this PR, and some UX issues when downloading files which may or may not matter to the stakeholders.  That being said, unless you think the UX issues are severe enough that you want to look at them _now_, I think we can merge this in.

**Comment by emetsger** [comment]
> @emetsger This was a fascinating PR. I think, reading the description, it makes sense to create issues for:
> 
> * review media types and suffix rules with stakeholders

Agree, happy to do this.

> * Verify consequences of File entities referring to same underlying bytestream (maybe something for vendor)

Also agree, and can create an issue.

> * Verify sufficiency of sha1 hashing

How best to do this?

> 
> I also noticed a few UX issues that ought to at least be investigated in separate issues, unless you think it makes sense to address in this PR. I'm not sure if stakeholders would care or not.
> 
> * Downloaded filenames do not have suffixes and correlate with the hash rather than filename from the metadata
> * MIME type when downloading files always seems to be application/octet-stream

Yes, both of these are particularly irritating.  The information is present in the `File` entity, so I'm not sure what is happening here.  I mean, it seems to me that Drupal could be using the File entity name and media type and sending `Content-Disposition`, but it just doesn't.  I'll create an issue to follow up especially to see if the behavior is different for interactively-created media.

(I suspect that Drupal simply punts on this, and relies on the browser handling by providing direct link.  If we want this behavior to change - which I think is a reasonable thing to do if we have cycles - we'd need to write some sort of file download controller or see if such a module already exists.  Alternately, we can revert the use of content addressed files)

**Comment by birkland** [comment]
> > * Verify sufficiency of sha1 hashing
> 
> How best to do this?

I think this would just be an issue to think through the consequences.  sha1 is [broken](https://shattered.io/), but it is unclear exactly what our "attack surface" is on that, what happens when colliding files are uploaded, and whether it matters.  

> it seems to me that Drupal could be using the File entity name and media type and sending `Content-Disposition`, but it just doesn't. I'll create an issue to follow up especially to see if the behavior is different for interactively-created media.
> (I suspect that Drupal simply punts on this, and relies on the browser handling by providing direct link. If we want this behavior to change - which I think is a reasonable thing to do if we have cycles - we'd need to write some sort of file download controller or see if such a module already exists. Alternately, we can revert the use of content addressed files)

Yeah.  I also wonder if it is really nginx calling the shots here, i.e. it's "just serving files in this one directory called sites/default/files".  If so, this would inherently change when we have se-backed files.  I have no idea whatsoever what would happen in the end, though.  It's possible this issue could go away if s3fs deals with file metadata explicitly.

**Comment by emetsger** [comment]
> Yeah. I also wonder if it is really nginx calling the shots here, i.e. it's "just serving files in this one directory called sites/default/files". If so, this would inherently change when we have se-backed files. I have no idea whatsoever what would happen in the end, though. It's possible this issue could go away if s3fs deals with file metadata explicitly.

Could be.  Either way, another consideration is that without any kind of download controller, direct links provide unauthenticated access to files.  Depending on what our APIs or feeds expose that may pose a risk with respect to copyrighted materials or high res images what we don't want to give out.
